plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/ptf/Query.java

```diff
@@ -0,0 +1,136 @@
+import io.trino.spi.function.table.Descriptor;
+import io.trino.spi.function.table.ScalarArgument;
+import io.trino.spi.function.table.ScalarArgumentSpecification;
+import io.trino.spi.function.table.TableFunctionAnalysis;
+import io.trino.spi.predicate.TupleDomain;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+
+import static io.trino.spi.function.table.ReturnTypeSpecification.GenericTable.GENERIC_TABLE;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class Query
+        implements Provider<ConnectorTableFunction>
+{
+    public static final String SCHEMA_NAME = "system";
+    public static final String NAME = "query";
+
+    private final PinotMetadata metadata;
+    private final PinotClient client;
+
+    private final PinotTypeConverter pinotTypeConverter;
+
+    @Inject
+    public Query(PinotMetadata metadata, PinotClient client, PinotTypeConverter pinotTypeConverter)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.client = requireNonNull(client, "client is null");
+        this.pinotTypeConverter = requireNonNull(pinotTypeConverter, "pinotTypeConverter is null");
+    }
+
+    @Override
+    public ConnectorTableFunction get()
+    {
+        return new QueryFunction(metadata, client, pinotTypeConverter);
+    }
+
+    public static class QueryFunction
+            extends AbstractConnectorTableFunction
+    {
+        private final PinotMetadata metadata;
+        private final PinotClient client;
+        private final PinotTypeConverter pinotTypeConverter;
+
+        public QueryFunction(PinotMetadata metadata, PinotClient client, PinotTypeConverter pinotTypeConverter)
+        {
+            super(
+                    SCHEMA_NAME,
+                    NAME,
+                    List.of(ScalarArgumentSpecification.builder()
+                            .name("QUERY")
+                            .type(VARCHAR)
+                            .build()),
+                    GENERIC_TABLE);
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.client = requireNonNull(client, "client is null");
+            this.pinotTypeConverter = requireNonNull(pinotTypeConverter, "pinotTypeConverter is null");
+        }
+
+        @Override
+        public TableFunctionAnalysis analyze(
+                ConnectorSession session,
+                ConnectorTransactionHandle transaction,
+                Map<String, Argument> arguments,
+                ConnectorAccessControl accessControl)
+        {
+            requireNonNull(arguments, "arguments is null");
+            String query = ((Slice) ((ScalarArgument) arguments.get("QUERY")).getValue()).toStringUtf8();
+            DynamicTable dynamicTable = DynamicTableBuilder.buildFromPql(metadata, query, client, pinotTypeConverter);
```

Comment:
This line may throw `org.apache.pinot.sql.parsers.SqlCompilationException` when the query is invalid. We should throw `TrinoException` instead.